### PR TITLE
Merge Dev to Master 

### PR DESCRIPTION
**Pull request by hoangthuc701**

### CODE REVIEW

File: client/src/components/AskQuestion/AskQuestion.jsx [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+	const user = JSON.parse(localStorage.getItem('profile'));
+	const [form, setForm] = useState(initialState);
+	const [body, setBody] = useState('');
+	const tags = useSelector((store) => store.tags?.recommendTags);
+	const [filteredTags, setFilteredTags] = useState([]);
+	const [errors, setError] = useState({ titleError: '', bodyError: '', tagError: '' });
+	const validate = (form, body) => {
+		let errors = { titleError: '', bodyError: '', tagError: '' };
+		if (form.title?.length <= 20) errors.titleError = 'Title must be at least 20 characters!';
+		if (body?.length < 20) errors.bodyError = 'Body must be at least 20 characters!';
```
Comment (hoangthuc701) [inline client/src/components/AskQuestion/AskQuestion.jsx:32]: we should not use magic number. we should create 1 constant variable and use them.

File: client/src/components/Auth/Signup.jsx [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+
+		if (formData.password === '') {
+			errors.passwordError = 'Password is required!';
+		} else if (formData.password.length < 8) {
+			errors.passwordError = 'Password length cannot be less than 8';
+		} else if (!formData.password.match(/[a-zA-Z]/)) {
+			errors.passwordError = 'Password must contain at least one alphabetical character';
+		} else if (!formData.password.match(/[0-9]/)) {
+			errors.passwordError = 'Password must contain at least one number';
+		} else if (formData.password === formData.email) {
```
Comment (hoangthuc701) [inline client/src/components/Auth/Signup.jsx:45]: We can be refactored by creating a valid password checker function and using it.

File: client/src/components/DetailQuestion/Detail.jsx [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+
+	const handleAddNewAnswerButton = () => {
+		setAnswerAttempt(!answerAttempt);
+	};
+
+	const validate = () => {
+		const errors = { bodyError: '' };
+		if (body?.length <= 20) {
+			errors.bodyError = 'Body must be at least 20 characters!';
+		}
```
Comment (hoangthuc701) [inline client/src/components/DetailQuestion/Detail.jsx:43]: Magic number. Please refactor it.

File: client/src/components/Header/AuthBtn.jsx [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+		width: '400px',
+	},
+	purple: {
+		color: theme.palette.getContrastText(deepPurple[500]),
+		backgroundColor: deepPurple[500],
+		marginRight: theme.spacing(2),
+	},
+}));
+const AuthBtn = () => {
+	const [user, setUser] = useState(JSON.parse(localStorage.getItem('profile')));
```
Comment (hoangthuc701) [inline client/src/components/Header/AuthBtn.jsx:36]: You can create a function to get user info and use it.  should not be called and retrieved directly from local storage. If you change the save location later, you will have to edit many places.

Consider other similar places.

File: client/src/components/Pagination/Pagination.jsx [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+
+import useStyle from './Pagination.style';
+
+const Pagination = () => {
+	const classes = useStyle();
+	return (
+		<div className={classes.pagination}>
+			<PaginationButton>&lsaquo;&lsaquo;</PaginationButton>
+			<PaginationButton>1</PaginationButton>
+			<PaginationButton>2</PaginationButton>
```
Comment (hoangthuc701) [inline client/src/components/Pagination/Pagination.jsx:13]: Why only from 1 to 3? we should reconsider this component,

File: client/src/index.js [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+import { createStore, applyMiddleware, compose } from 'redux';
+import thunk from 'redux-thunk';
+
+import { reducers } from './reducers';
+import App from './App';
+
+const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;
+
+const store = createStore(reducers, composeEnhancers(applyMiddleware(thunk)));
+
```
Comment (hoangthuc701) [inline client/src/index.js:13]: can refactor a function to return a store.

File: client/src/reducers/auth.js [hunk 7]
```diff
@@ -0,0 +1,16 @@
+import { SIGN_IN, SIGN_OUT } from '../constants/index';
+
+const authReducer = (state = { authData: null }, action) => {
+	switch (action.type) {
+		case SIGN_IN:
+			localStorage.setItem('profile', JSON.stringify(action?.payload));
```
Comment (hoangthuc701) [inline client/src/reducers/auth.js:6]: We should create a function to save data.

File: client/src/reducers/auth.js [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+			return { ...state, authData: action?.payload };
+		case SIGN_OUT:
+			localStorage.removeItem('profile');
```
Comment (hoangthuc701) [inline client/src/reducers/auth.js:9]: You can create a class to add, remove, and edit data in local storage.

File: server/controllers/auth.js [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+	try {
+		const decodedToken = jwt.verify(queryToken, SECRET_KEY);
+
+		if (decodedToken) {
+			const { email } = decodedToken;
+			const user = await User.findOne({ email });
+
+			if (!user) return res.status(400).json({ message: 'This link is invalid' });
+
+			res.status(200).redirect(`http://localhost:3000/auth/reset-password/${user._id}`);
```
Comment (hoangthuc701) [inline server/controllers/auth.js:103]: We should create an environment variable and use it.

File: server/controllers/auth.js [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+		res.status(500).json({ message: 'Something went wrong' });
+	}
+};
+
+export const changePassword = async (req, res) => {
+	const { newPassword } = req.body;
+	const { id } = req.params;
+	const hashedPassword = await bcrypt.hash(newPassword, 12);
+	await User.findByIdAndUpdate({ _id: id }, { password: hashedPassword, dateUpdated: new Date() }, { new: true });
+	res.status(200).redirect('http://localhost:3000/auth/signin');
```
Comment (hoangthuc701) [inline server/controllers/auth.js:115]: We should create an environment variable and use it.

File: server/routes/question.js [hunk 11]
```diff
@@ -0,0 +1,13 @@
+import express from 'express';
+
+import * as controller from '../controllers/question.js';
+
+const router = express.Router();
+
+router.post('/', controller.createQuestion);
```
Comment (hoangthuc701) [inline server/routes/question.js:7]: duplicate code

File: client/src/components/Pagination/Pagination.jsx (hunk 5 above)
Comment (huyhuynh2705) [reply client/src/components/Pagination/Pagination.jsx:13]: We haven't done this component yet, we will complete it this week.

**Pull request merged**